cherry-pick 2.0: sql: fix bug and panic with INSERT/UPSERT

knz opened the pull request:
Picks #23657. (split UPSERT from INSERT, split fast path from slow path).

This fixes a panic and a bug.

Dependencies:

Picks #22784. (ivarHelper cleanup).
Picks #23526. (code comments / doc only, but touches the insert/upsert code).

RaduBerinde committed:
sql: replace IndexedVarHelper in SemaContext with IndexedVarContainer

The `SemaContext.IVarHelper` felt strange to me (why would we want a
helper in a semantic context?) so I looked at how it is used. Turns
out we only use it for its container (which makes sense). This leads
to situations where we create `IndexedVarHelper`s just to populate
this field.

We change the field to be an `IndexedVarContainer`.

Release note: None

RaduBerinde committed:
sql: replace IndexedVarHelper in EvalContext with IndexedVarContainer

`EvalContext.IndexedVarHelper` is only used for its container. In many
contexts we create a helper just to populate this field.

Changing the field to `IndexedVarContainer`.

Release note: None

knz committed:
sql: add missing comments; increase code clarity

Release note: none

knz committed:
sql: split the code for UPSERT from the INSERT code

Prior to this patch an attempt was made to share local execution logic
between UPSERT and INSERT.

There is however sufficient complexity in UPSERT that it drowns the
commonalities with INSERT. As a result, the INSERT logic is hard to
comprehend and meanwhile is not much maintainability gain from DRY.

To clarify the INSERT code, and prepare it for an upcoming fix that
gives it proper batching behavior, the execution logic for UPSERT is
removed from it and segregated into a new upsertNode.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23658)
<!-- Reviewable:end -->

knz committed:
sql: split the code for general and fast path of UPSERT

The tableWriter code for the general case and the fast path in UPSERT
is rather different. There is no benefit to having them in the same
struct/object.

This patch separates them into two structs to make the code easier to
understand.

Also, adding lots of comments and fixing a couple of bugs (see release
notes below).

Release note (bug fix): fixed a bug where INSERT ... ON CONFLICT DO
UPDATE would accept to update the same column twice.

Release note (bug fix): fixed a panic when a query incorrectly would
attempt to use an aggregation or window function in ON CONFLICT DO
UPDATE.

knz commented:
Hi Radu, could you review this cherry-pick. I'm picking you relatively arbitrarily, but not the least because you authored the first two commits that this picks, as a dependency to the main one (the last, which fixes a bug).

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Very nice improvements to the code and comments.

---

Review status: 0 of 40 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23658#-:-L7HVXlxmGvUUE0OEDV5:b-gzxoa1)*
<!-- Sent from Reviewable.io -->

knz commented:
Thank you!
